amundsen_common/models/user.py

```diff
@@ -1,29 +1,85 @@
-from typing import Optional
+from typing import Optional, Dict
 
 import attr
+from marshmallow import ValidationError, validates_schema, pre_load
 from marshmallow_annotations.ext.attrs import AttrsSchema
 
+"""
+TODO: Explore all internationalization use cases and
+redesign how User handles names
+
+TODO - Delete pre processing of the Data
+Once all of the upstream services provide a complete User object we will no
+longer need to supplement the User objects as done in `preprocess_data`
+"""
+
 
 @attr.s(auto_attribs=True, kw_only=True)
 class User:
-    # ToDo (Verdan): Make ID a required field.
+    # ToDo (Verdan): Make user_id a required field.
     #  In case if there is only email, id could be email.
     #  All the transactions and communication will be handled by ID
-    id: Optional[str] = None
+    user_id: Optional[str] = None
     email: Optional[str] = None
     first_name: Optional[str] = None
     last_name: Optional[str] = None
     full_name: Optional[str] = None
+    display_name: Optional[str] = None
     is_active: bool = True
     github_username: Optional[str] = None
     team_name: Optional[str] = None
     slack_id: Optional[str] = None
     employee_type: Optional[str] = None
     manager_fullname: Optional[str] = None
+    manager_email: Optional[str] = None
+    manager_id: Optional[str] = None
     role_name: Optional[str] = None
+    profile_url: Optional[str] = None
+    # TODO: Add frequent_used, bookmarked, & owned resources
 
 
 class UserSchema(AttrsSchema):
     class Meta:
         target = User
         register_as_scheme = True
+
+    # noinspection PyMethodMayBeStatic
+    def _str_no_value(self, s: Optional[str]) -> bool:
+        # Returns True if the given string is None or empty
+        if not s:
+            return True
+        if len(s.strip()) == 0:
+            return True
+        return False
+
+    @pre_load
+    def preprocess_data(self, data: Dict) -> Dict:
+        if self._str_no_value(data.get('user_id')):
+            data['user_id'] = data.get('email')
+
+        if self._str_no_value(data.get('profile_url')):
+            data['profile_url'] = ''
+            if data.get('GET_PROFILE_URL'):
```

Comment:
How about having `GET_PROFILE_URL` as constant?

Reply:
@jinhyukchang this is a config variable/function in frontend repository. and to make it backward compatible passing it through the frontend service. Also, I think we should try avoid dividing the config variables across different services as much as possible, and keep them in one place for easy customization, that's why did not move this config variable from frontend service.

Reply:
@verdan One way around this in the future would be if the user's `profile_url` became a piece of User metadata to link to the internal profile, in the same way that we have `dashboard_url` for Dashboard metadata to link to the internal source. 

We originally created this configuration because it was a quick frontend-only workaround to support the custom employee profile link feature when we implemented Amundsen People.

Reply:
@jinhyukchang @feng-tao I believe we're having issues with this when trying to deploy internally. Features related to the user schema are broken and the error we get is `TypeError: Object of type 'function' is not JSON serializable`, which I believe is coming from this code because `'GET_PROFILE_URL'` is a function. 

Any ideas on how to work around this, and ultimately whether or not we can have `profile_url` as user metadata in the future to be able to delete this code?

Reply:
hey @ttannis let me take a look as well, but somehow I recall @verdan 's pr just directly copy the model from FE into common (e.g https://github.com/lyft/amundsenfrontendlibrary/pull/440/files) . Any reason why it suddenly breaks now and not before?

Reply:
actually there is small difference...

Reply:
I believe @ttanis is right. And I made a mistake here. One of the solution to solve this is to generate the profile url in frontend after the user dump.

But again, depends on whether we want this to be a part of user metadata and if we need this config here.

Reply:
Thanks all. Filed the follow up: https://github.com/lyft/amundsen/issues/456